회원가입 API 개발

zlzzlzz2l opened the pull request:
# 회원가입 API 개발

회원가입과 이메일, 닉네임 중복 기능을 추가했습니다.

## 회원가입

모든 정보를 입력했을 때
![image](https://user-images.githubusercontent.com/48669011/145711516-72884649-daba-4e8e-8471-cbb8632ce715.png)
![image](https://user-images.githubusercontent.com/48669011/145711526-68f73c90-cbdb-4aba-8466-34522c791a06.png)

사용자 직업, 생일, 성별을 입력 안했을 때
![image](https://user-images.githubusercontent.com/48669011/145711468-11fccca4-3495-4e14-b05b-a93dc087d931.png)
![image](https://user-images.githubusercontent.com/48669011/145711478-c7572cbe-3bab-4ee2-9b17-7fd12f89c804.png)

회원가입 유효성 검사
![image](https://user-images.githubusercontent.com/48669011/145711426-c7ce5494-0738-490f-8923-bc2dfc469a41.png)

### 회원가입 API
![image](https://user-images.githubusercontent.com/48669011/145711196-ca226f4a-f6ad-48c6-835f-aacd913c6043.png)
![image](https://user-images.githubusercontent.com/48669011/145711221-aaf86977-5a01-4d4c-984a-9adffbd71a97.png)
![image](https://user-images.githubusercontent.com/48669011/145711227-41f7c831-eba5-4d5c-b697-0647686f59e2.png)
![image](https://user-images.githubusercontent.com/48669011/145711248-206385f8-dcf2-46b8-a0db-4e69ea704c2f.png)
![image](https://user-images.githubusercontent.com/48669011/145711264-26231a62-67ab-49ad-8f2a-d2595765c7d5.png)
![image](https://user-images.githubusercontent.com/48669011/145711266-c5100fd9-161d-41d7-be66-9a04aad0d7f3.png)

zlzzlzz2l committed:
Chore: #7 - Validation 라이브러리 추가

zlzzlzz2l committed:
Update: #7 - 회원 Dao 회원가입, 이메일 & 닉네임 중복 체크 메서드 선언부 추가

zlzzlzz2l committed:
Update: #7 - 회원 Dao 회원가입 메서드 구현

zlzzlzz2l committed:
Update: #7 - 회원 Dao 이메일 중복 체크 구현

zlzzlzz2l committed:
Update: #7 - 회원 Dao 닉네임 중복 체크 구현

zlzzlzz2l committed:
Update: #7 - 회원 Dto 유효성 체크 추가

zlzzlzz2l committed:
Update: #7 - 회원 서비스 회원가입, 이메일 & 닉네임 중복 체크 메서드 선언부 추가

zlzzlzz2l committed:
Update: #7 - 회원 서비스 회원가입 메서드 구현

zlzzlzz2l committed:
Update: #7 - 회원 서비스 이메일 중복 체크 메서드 구현

zlzzlzz2l committed:
Update: #7 - 회원 서비스 닉네임 중복 체크 메서드 구현

zlzzlzz2l committed:
Update: #7 - 회원 컨트롤러 회원가입, 이메일 & 닉네임 메서드 추가

zlzzlzz2l committed:
Create: #7 - Validation 에러 메시지 JSON 직렬화 메서드 구현

yjs1213 reviewed (approved):
`SELECT`, `WHERE`, `VALUES` 등의 MySQL 키워드를 가독성을 위해 대문자로 바꾸는건 어떤가요?

물론 필수는 아닙니다!!

참고 : https://www.codeit.kr/community/threads/26154

0xe82de commented on src/main/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImpl.java:
```diff
@@ -89,4 +89,22 @@ public void insertUser(UserDto user) {
+                .append(TABLE_NAME).append(" ")
+                .append("where user_email = ?");
+
+        List<String> resultEmail = jdbcTemplate.query(sql.toString(), new RowMapper<String>() {
```
반환 타입을 String으로 바꿔도 좋을 것 같습니다. 바꾸게 되면 메서드 반화타입과 서비스 코드에서 리턴받는 변수 타입도 바꿔야 합니다.

zlzzlzz2l replied:
반영합니다.

0xe82de commented on src/main/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImpl.java:
```diff
@@ -107,4 +107,20 @@ public String mapRow(ResultSet rs, int rowNum) throws SQLException {
+                .append("select user_nickname from").append(" ")
+                .append(TABLE_NAME).append(" ")
+                .append("where user_nickname = ?");
+        List<String> resultNickname = jdbcTemplate.query(sql.toString(), new RowMapper<String>() {
```
이 부분도 반환타입을 String으로 받아도 좋을 것 같습니다.

zlzzlzz2l replied:
반영합니다.

zlzzlzz2l committed:
Style: #7 - 회원 Dao 쿼리문 수정

가독성을 위해 소문자에서 대문자로 변경했습니다.

zlzzlzz2l committed:
Style: #7 - 회원 컨트롤러 이메일 중복 체크 메서드명 수정

스펠링 누락으로 emailDulication 메소드를 emailDuplication 메소드로 변경했습니다.

zlzzlzz2l committed:
Style: #7 - 회원 서비스 이메일 중복 체크 메서드명 수정

스펠링 누락으로 emailDulication 메소드를 emailDuplication 메소드로 변경했습니다.

zlzzlzz2l committed:
Refactor: #7 - 회원 Dao 이메일 & 닉네임 중복 체크 메서드 타입 변경

회원 Dao 이메일 & 닉네임 중복 체크 메서드 반환 타입을 String으로 바꿨기 때문에 선언부에서도 변경했습니다.

zlzzlzz2l committed:
Style: #7 - 회원 서비스 이메일 중복 체크 메서드명 수정

스펠링 누락으로 emailDulication 메서드를 emailDuplication 메서드로 변경했습니다.

zlzzlzz2l committed:
Refactor: #7 - 회원 Dao 이메일 & 닉네임 중복 체크 메서드 타입 변경

이메일과 닉네임은 중복될 수 없고, 결과 행이 무조건 하나이기 때문에 queryForObject를 사용
이메일 & 닉네임 중복 체크 메서드 타입을 String으로 변경
중복되지 않은 이메일과 닉네임이 들어오면 EmptyResultDataAccessException 예외 발생
예외가 발생하면 null을 반환하기 위해 try-catch를 사용

zlzzlzz2l committed:
Refactor: #7 - 회원 서비스 이메일 & 닉네임 중복 체크 메서드 타입 변경

회원 Dao 이메일 & 닉네임 메서드의 반환 타입이 String이므로 회원 서비스도 변경

zlzzlzz2l committed:
Refactor: #7 - 회원 컨트롤러 이메일 & 닉네임 중복 체크 메서드명 수정

zlzzlzz2l committed:
Refactor: #7 - 회원 서비스 이메일 & 닉네임 중복 체크 메서드명 수정

zlzzlzz2l committed:
Refactor: #7 - 회원 서비스 이메일 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #7 - 회원 서비스 닉네임 중복 체크 메서드 수정

0xe82de reviewed (approved):
고생하셨습니다~!

yjs1213 reviewed (approved):
고생하셨어요~
